numpy/lib/npyio.py

```diff
@@ -1908,16 +2039,43 @@ def genfromtxt(fname, dtype=float, comments='#', delimiter=None,
         column_types = [conv.type for conv in converters]
         # Find the columns with strings...
         strcolidx = [i for (i, v) in enumerate(column_types)
-                     if v in (type('S'), np.string_)]
+                     if v == np.unicode_]
+
+        typestr = 'U'
+        if byte_converters and strcolidx:
+            # convert strings back to bytes for backward compatibility
+            warnings.warn(
+                "Reading strings without specifying the encoding argument is "
+                "deprecated. Set encoding, use None for the system default.",
+                np.VisibleDeprecationWarning, stacklevel=2)
+            try:
+                for j in range(len(data)):
+                    row = list(data[j])
+                    for i in strcolidx:
+                        row[i] = row[i].encode('latin1')
+                    data[j] = tuple(row)
+                typestr = 'S'
+            except UnicodeEncodeError:
+                # we must use unicode, revert encoding
+                for k in range(0, j + 1):
+                    row = list(data[k])
+                    for i in strcolidx:
+                        if isinstance(row[i], bytes):
+                            row[i] = row[i].decode('latin1')
+                    data[k] = tuple(row)
```

Comment:
IMO this block would be clearer as:
```
try:
    new_data = []
    for row_tup in data):
        row = list(row_tup)
        for i in strcolidx:
            row[i] = row[i].encode('latin1')
        row_tup = tuple(row)
        new_data.append(row_tup)
except UnicodeEncodeError:
     # we must use unicode
else:
    data = new_data
```

Or maybe even

```
def encode_unicode_cols(row_tup):
    row = list(row_tup)
    for i in strcolidx:
        row[i] = row[i].encode('latin1')
    return tuple(row)

try:
    data = [encode_unicode_cols(r) for r in data]
except UnicodeEncodeError:
    pass
```

Reply:
`type_str` needs to be a numpy type, 'U' or 'S'. The 'U' is the default here, overridden if the `try` succeeds.